proposals/compile-time-dependency-resolution.md

```diff
@@ -0,0 +1,590 @@
+fetchById<User>(11829, UserRepository()) // you can provide it manually.
+fetchById<User>(11829, repository=UserRepository()) // you can provide it manually with named application.
+```
+
+When `with` is used in class constructors, it is important to **add val to extension class fields** to make sure they are accessible in the scope of the class. Here, the `with` keyword adds the value to the scope of every method in the class. To showcase this, let's say we have a `Validator<A>`, like:
+
+```kotlin
+interface Validator<A> {
+  fun A.isValid(): Boolean
+}
+```
+
+In this scenario, the following classes would be equivalent:
+
+```kotlin
+// Kotlin + KEEP-87
+extension class ValidatedRepository<A>(with val V: Validator<A>) : Repository<A> {
+
+    val storedUsers: MutableMap<Int, A> = mutableMapOf() // e.g. users stored in a DB
+
+    override fun loadAll(): Map<Int, A> {
+        return storedUsers.filter { it.value.isValid() }
+    }
+
+    override fun loadById(id: Int): A? {
+        return storedUsers[id]?.let { if (it.isValid()) it else null }
+    }
+
+    override fun A.save() {
+        storedUsers[generateKey(this)] = this
+    }
+}
+
+// Regular Kotlin
+class ValidatedRepository<A>(val V: Validator<A>) : Repository<A> {
+
+    val storedUsers: MutableMap<Int, A> = mutableMapOf() // e.g. users stored in a DB
+
+    override fun loadAll(): Map<Int, A> {
+        with (V) {
+            return storedUsers.filter { it.value.isValid() }
+        }
+    }
+
+    override fun loadById(id: Int): A? {
+        with (V) {
+            return storedUsers[id]?.let { if (it.isValid()) it else null }
+        }
+    }
+
+    override fun A.save() {
+        storedUsers[generateKey(this)] = this
+    }
+}
+```
+
+As you can see in the first example, `A.isValid()` becomes available inside the method’s scope automatically. The equivalent version of doing this without KEEP-87 is to manually add `with (V)` inside each method, as you can see in the second example.  
+
+## Composition and chain of evidences
+
+Constraint interface declarations and extension evidences can encode further constraints on their type parameters so that they can be composed nicely:
+
+```kotlin
+package com.data.instances
+
+import com.data.Repository
+import com.domain.User
+
+data class Group<A>(val values: List<A>)
+
+extension class GroupRepository<A>(with val repoA: Repository<A>) : Repository<Group<A>> {
+  override fun loadAll(): Map<Int, Group<A>> {
+    return repoA.loadAll().mapValues { Group(it.value) }
+  }
+
+  override fun loadById(id: Int): Group<A>? {
+    return repoA.loadById(id)?.let { Group(it) }
+  }
+
+  override fun Group<A>.save() {
+    this.items.map { repoA.run { it.save() } }
+  }
+}
+```
+
+The above extension provides evidence of a `Repository<Group<A>>` as long as there is a `Repository<A>` in scope. The Call site would look like:
+
+```kotlin
+fun <A> fetchGroup(with repo: GroupRepository<A>) = loadAll()
```

Comment:
Shouldn't the argument be `with repo: Repository<Group<A>>`? I.e. it should be a requirement of *some* implementation of the given type, not the specific one given above.

Reply:
@raulraja this seems correct

Reply:
I'm focusing right now on something more pressing in arrow but feel free to push a change or PR for all the suggested fixes. Thanks